Mostly type-hinted helper classes for inclusion in v2

paul-griffith opened the pull request:
###  Summary

This is a direct copy-paste of the pseudo-library that wraps `python-slackclient` I use internally. The big advantage is obviously typed interactions - especially the newer stuff (ie, blockkit) has pretty solid type hinting throughout. 

I'm offering up this PR without any adjustments having been made - if it's something y'all are interested in, I could potentially spend some time making changes to better fit it into this library. Happy to review with anybody - I'm also in the dev4slack group semi-often if you want to reach me realtime.


### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/python-slackclient/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

RodneyU215 commented:
@paul-griffith This is awesome! I'm still reading through it, but I'm very interested in working together to get this into a mergeable state for v2.

1oglop1 commented:
This is super nice! it would help me a lot!

DandyDev commented:
It would be awesome for this to be merged in. What needs to be done for this PR to be accepted?

RodneyU215 commented:
There are currently a few high priority bugs I'm working on so it's taken me a while to get to this. This is definitely still a PR we'd like to see merged! Right now the biggest blocker is Unit tests and documentation.

DandyDev commented:
The question is if you want to merge this as-is, with the [extra wrapper](slack/web/objects/wrapper.py), or if you maybe want to integrate the helper classes with the existing clients. I'd say the latter is preferable, but that requires some extra work which I don't know if @paul-griffith wants to put in.

RodneyU215 commented:
No, the [extra wrapper](https://github.com/slackapi/python-slackclient/pull/400/commits/ed8d89392bbc8bec0dfb6a4735c5372a581ba26c#diff-1b520c2c2bb80a5733e5fb39350a0d43) is an area that I'd like to discuss more. Preferably, I'd love to work to get the message builder classes merged first. This could be used immediately with the existing API methods.

RodneyU215 commented:
@paul-griffith I want to thank you again for contributing such a phenomenal PR. It's clear you've put a great deal of work into this and it's going to be provide a lot of value to the community.

DandyDev commented:
If I can help put this into the desired shape, let me know @paul-griffith @RodneyU215. This addition would greatly simplify the code of [my own bot framework](https://github.com/DandyDev/slack-machine)

paul-griffith commented:
I've made improvements and started to add tests to my own lib since I
opened this, but before I do more specific work for integrating I want to
double check with my own boss about how much time I can dedicate to this. I
also just updated my own codebase to work with v2 of this package.

Rodney, may be worth arranging a time we can have a quick discussion
real-time to see what you would want changed, also - that would help me see
how much time would be needed.

On Tue, Jun 4, 2019, 10:52 AM Daan Debie <notifications@github.com> wrote:

> If I can help put this into the desired shape, let me know @paul-griffith
> <https://github.com/paul-griffith> @RodneyU215
> <https://github.com/RodneyU215>. This addition would greatly simplify the
> code of my own bot framework <https://github.com/DandyDev/slack-machine>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/slackapi/python-slackclient/pull/400?email_source=notifications&email_token=AJMFY3RYAJZSWSDGXHJYY3DPY2TVTA5CNFSM4HIPHII2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODW5LNEQ#issuecomment-498775698>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJMFY3UAXMEKRZZZEWTSZGLPY2TVTANCNFSM4HIPHIIQ>
> .
>

RodneyU215 commented:
@paul-griffith Happy to chat this week. I've sent you a DM in dev4slack and you can also email me at `rodney@slack-corp.com` so we can coordinate with a calendar invite.

paul-griffith commented on slack/web/classes/dialogs.py:
```diff
@@ -0,0 +1,453 @@
+        self._notify_on_cancel = False
+        self._state = None
+
+    def title(self, title: str) -> "DialogBuilder":
```
This annotation could be skipped, _or_ could `from future import annotations` in order to drop the string casing, although that does have performance implications on 3.6 I believe

paul-griffith commented on slack/web/classes/actions.py:
```diff
@@ -0,0 +1,207 @@
+class Action(JsonObject):
+    attributes = {"name", "text", "type", "url"}
+
+    def __init__(self, *, text: str, type: str, name: str = None, url: str = None):
```
I followed the larger lead of the library here and switched everything to require kwargs - not sure if that's something we want these classes to do or not

paul-griffith commented on slack/web/classes/objects.py:
```diff
@@ -0,0 +1,335 @@
+OptionTypes = (Option, OptionGroup)
+
+
+def extract_json(
```
Not sure how you feel about this method - if you scan usages, it's mostly just removing boilerplate to allow us to be nicer about letting someone attach a dictionary directly where we would normally expect one of these objects

paul-griffith commented:
I don't plan on significantly changing anything in the core lib at this point - just need to add test coverage at this point. Happy for anyone else to contribute tests, or even just known-failing-cases - coverage right now is basically limited to whatever is explicitly not allowed in the API documentation, but I'm confident there's plenty of scenarios that aren't covered by the JSON validation this currently has.

paul-griffith commented on slack/web/classes/actions.py:
```diff
@@ -0,0 +1,301 @@
+
+        https://api.slack.com/docs/message-buttons
+
+        :param name: Name this specific action. The name will be returned to your
```
Tried to follow a template for docstrings:
> basic explanation
> link to API documentation, if I could find a useful permalink
> param explanation. If possible/relevant, copied directly from API reference page(s)

paul-griffith replied:
Done! I was honestly just using PyCharm's default :)

paul-griffith commented on slack/web/classes/attachments.py:
```diff
@@ -0,0 +1,350 @@
+        footer_icon: str = None,
+        ts: int = None,
+    ):
+        """
```
The docstring/param list on these two classes is a bit ridiculous, but there's not really any meta-information to keep track of that helps make a builder class worthwhile, IMO

paul-griffith replied:
On consideration, I _could_ see some utility to an AttachmentBuilder - something like:
```
AttachmentBuilder()
  .text("Some attachment")
  .footer("footer", icon="icon url")
  .author("jim bob", link="jimbob.com", icon="jimbob.com/picture.jpg")
  .field("*Score*", "20")
  .field("*Series*", "Some book series")
  .linkbutton("button", "Click me!", "http://google.com")
  .user_selector()
  .callback_id("abcdefg")
  .build()
```

But, that also ties into the point that pure attachments shouldn't really be encouraged - although I do hope either blocks get a color attribute soon or BlockAttachments never go away :crossed_fingers:

DandyDev replied:
I like the `AttachementBuilder`!

RodneyU215 replied:
While I agree an `AttachementBuilder` would be nice, I would not recommend investing too much time into it. Block Kit is the focus and will continue to see new features added.

paul-griffith commented on slack/web/classes/blocks.py:
```diff
@@ -0,0 +1,189 @@
+)
+
+
+class Block(JsonObject):
```
We had talked about a `BlockBuilder` or something along those lines, but when I started prototyping one the usage was pretty awkward - there's just too many combinations of blocks + elements + objects for a builder to really make sense, IMO

RodneyU215 commented on slack/web/classes/dialogs.py:
```diff
@@ -0,0 +1,890 @@
+
+        :param label: Label displayed to user. Required. No more than 48 characters.
+
+        :param min_query_length:  	Specify the number of characters that must be
```
I believe this is a typo tab.

RodneyU215 commented on slack/web/classes/interactions.py:
```diff
@@ -0,0 +1,123 @@
+        self.event_type = event["type"]
+        self.message_ts = event["message"]["ts"]
+        self.trigger_id = event["trigger_id"]
+        action = event["actions"][0]
```
Can you add a comment why the first action is singularly pulled out?

RodneyU215 commented on slack/web/classes/interactions.py:
```diff
@@ -0,0 +1,123 @@
+
+    def __init__(self, event: dict):
+        """
+        Convenience class to parse a dialog interaction payload from the events API
```
These convenience classes are definitely handy. Once https://github.com/slackapi/python-slackclient/issues/403 is picked up I'd imagine we could bake these in directly.

RodneyU215 commented on slack/web/classes/messages.py:
```diff
@@ -0,0 +1,62 @@
+    def attachments_length(self):
+        return self.attachments is None or len(self.attachments) <= 100
+
+    def get_json(self):
```
Minor nit:  Add a return type. `-> dict`

RodneyU215 commented on slack/web/classes/objects.py:
```diff
@@ -0,0 +1,522 @@
+
+    def get_json(self, option_type: str = "block") -> dict:
+        if option_type == "action":
+            # deliberately skipping JSON validators here - can't find documentation
```
Note for myself: Check on limits described here.

RodneyU215 commented:
@DandyDev If you could spare a moment to review this PR I'd love your feedback on this as well.

DandyDev commented:
@RodneyU215 I will have a look at it tonight (GMT+2 time). Happy to help!

DandyDev commented on slack/web/classes/dialogs.py:
```diff
@@ -0,0 +1,868 @@
+from abc import ABCMeta, abstractmethod
```
This module seems quite long. Can it be broken up somehow?

DandyDev commented on slack/web/classes/elements.py:
```diff
@@ -0,0 +1,557 @@
+import random
```
Also quite a long module...

DandyDev commented on slack/web/classes/objects.py:
```diff
@@ -0,0 +1,530 @@
+from datetime import datetime
```
Long module. You could maybe split the `JsonXXX` stuff from the rest?

DandyDev commented on slack/web/classes/messages.py:
```diff
@@ -0,0 +1,68 @@
+    def get_json(self) -> dict:
+        self.validate_json()
+        if len(self.text) > 40000:
+            logging.getLogger(__name__).error(
```
Maybe it would be better to define the logger at the top of the module, so it doesn't need to be created both here and on line 58?

DandyDev commented on slack/web/classes/blocks.py:
```diff
@@ -0,0 +1,199 @@
+    block_id_max_length = 255
+
+    def __init__(self, *, type: str, block_id: str = None):
+        self.type = type
```
`type` shadows a builtin, so maybe give the parameter another name?

DandyDev commented:
> @DandyDev If you could spare a moment to review this PR I'd love your feedback on this as well.

Done!

paul-griffith commented:
Solid points, @DandyDev. I moved the dialog sub-elements into a `dialog_elements.py` - I feel better about that so that users are "encouraged" to use the builder via `slack.web.classes.dialogs`, and have to go out of their way to pull in other classes. I also moved the common JSON handling into `__init__.py` for the package.

Agreed on `get_json`, but I'm just as unsure on a better replacement. `to_dict` or `to_object` might be better?

Also fixed the use of `Optional` and shadowing `type` when appropriate.

`elements.py` _is_ fairly long (with the latest commit, it's the largest file in this directory) but I can't see a good way to split it up like what was possible with `dialogs`. Open to suggestions.

DandyDev commented:
> Solid points, @DandyDev. I moved the dialog sub-elements into a `dialog_elements.py` - I feel better about that so that users are "encouraged" to use the builder via `slack.web.classes.dialogs`, and have to go out of their way to pull in other classes. I also moved the common JSON handling into `__init__.py` for the package.
> ...
> Also fixed the use of `Optional` and shadowing `type` when appropriate.

Great!

> Agreed on `get_json`, but I'm just as unsure on a better replacement. `to_dict` or `to_object` might be better?

I like `to_dict`!


> `elements.py` _is_ fairly long (with the latest commit, it's the largest file in this directory) but I can't see a good way to split it up like what was possible with `dialogs`. Open to suggestions.

I see a couple of element categories that could be split out:
- Select-related stuff
- Menu-related stuff

That's the only split I can think of.

paul-griffith commented:
One other option with `elements.py` would be moving abstract/base classes to a subfile, so that `from slack.web.classes.elements import <x>` stays pretty clean for actual use classes, but that feels low value to me.

DandyDev commented:
I'd say, leave it for now. For me it's a minor nitpick. If stuff is added to `elements.py` in the future, it can be reevaluated.

paul-griffith commented:
Okay, @RodneyU215, I'll defer to you on decisions about:
1. `get_json()` vs `to_dict` vs ???
2. `AttachmentBuilder`, yea or nay?
3. Refactor `elements.py`?

Otherwise, I think this is almost ready to merge? :crossed_fingers:

RodneyU215 commented:
@DandyDev Thanks for the review! 

@paul-griffith 
1. I agree `to_dict()` is a much better fit.
2. `AttachmentBuilder`, nay. I'd prefer not to invest too much into something we're actively working to replace.
3. `elements.py` is long, but the context is where it needs to be. Let's keep this as-is for now.

Thank you both for the progress on this. We're very close to getting this merged. Once the rename is complete let's get this shipped.
